Add option accept or reject channel requests

ViktorTigerstrom opened the pull request:
Adds support for manually accepting or rejecting a request to open a channel. Issue #1242

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4110,6 +4157,24 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
+		return self.internal_create_requested_channel(counterparty_node_id, their_features, msg);
+	}
+
+	fn internal_create_requested_channel(&self, counterparty_node_id: &PublicKey, their_features: InitFeatures, msg: &msgs::OpenChannel) -> Result<(), MsgHandleErrInternal>{
```
Instead of splitting the function here, we can simplify things a bit by splitting the function a few lines down and moving only the `SendAcceptChannel` message generation out. There's no harm in having a channel be inserted into the `channel_state.by_id` map even if its not accepted yet, we just have to make sure we reject a `FundingCreated` message in `Channel::funding_created` if the sender is trying to be too clever and skip waiting for our accept. This will require a new flag in the `Channel` which is updated when we (eventually) call `get_accept_channel` and checked in `funding_created`.

From there, `Event::OpenChannelRequest` can be tweaked to just include relevant info instead of doing it via a struct, and `respond_to_open_channel_request` can just take the `temporary_channel_id`. All that said, I'd suggest we err on the side of including too much data rather than too little, as the node operator may need extra data to decide if they want to accept a channel.

ViktorTigerstrom replied:
Thanks a lot for the detailed input! I'll start working on implementing your suggestions. 

Just to clarify though, how much/which data is enough for node operator to aid them in the decision to accept or reject the channel? Including all fields in the `OpenChannel` msg is too excessive, but given that I'd really appreciate if I'd could get some feedback on which fields should be included instead of selecting myself.

TheBlueMatt replied:
I'm not sure there's necessarily a clear answer, but maybe include, for now (we can always add more later): counterparty node id, chain_hash, temporary_channel_id, funding_satoshis, push_msat?

ViktorTigerstrom replied:
Thanks, I'll start with those.

TheBlueMatt commented on lightning/src/util/events.rs:
```diff
@@ -515,6 +535,12 @@ impl Writeable for Event {
 				})
 			},
+			&Event::OpenChannelRequest { ref open_channel_request_info } => {
+				11u8.write(writer)?;
```
11 is already used for `DiscardFunding`. We need a new value here. That said, we shouldn't actually ever load this object from disk, similar to how we handle `PendingHTLCsForwardable` and `FundingGenerationReady` - if we have such an object when we go to serialize, it won't make sense when we go to reload it. In lightning, if a channel has not exchanged funding messages, it is dropped on reconnect (ie restart), so the peer will have forgotten about the channel entirely anyway.

ViktorTigerstrom replied:
Thanks for the feedback! I'll make sure to fix that.

ViktorTigerstrom commented:
I've now updated the PR to hopefully resolve the issues raised in the feedback. I can of course squash the 2 commits into 1 if that's more suitable.

TheBlueMatt commented on lightning/src/util/events.rs:
```diff
@@ -515,6 +540,11 @@ impl Writeable for Event {
 				})
 			},
+			&Event::OpenChannelRequest { .. } => {
+				16u8.write(writer)?;
```
This needs to be odd - our read behavior is "even, and we dont understand it -> assume we cant read anything and it is an important event, refuse to deserialize at all", but here its an unimportant (so unimportant we *want* to throw it away) event, so we should use an odd value to indicate that. We should also add a comment on the read-side code noting that the value we use is used for `OpenChannelRequest` (maybe a match branch).

ViktorTigerstrom replied:
Thanks again for the detailed input! I'll assign it to 17 instead.

TheBlueMatt commented on lightning/src/util/events.rs:
```diff
@@ -403,6 +403,31 @@ pub enum Event {
+		chain_hash: BlockHash,
+		/// The channel value of the requested channel.
+		funding_satoshis: u64,
+		/// The amount to push to the counterparty if the channel request is accepted, in milli-satoshi.
```
That's us! Maybe say something like "the amount the counterparty will give us if the channel is opened" or talk about it as "our starting balance"

ViktorTigerstrom replied:
Oh my bad, should have known that!

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4101,6 +4101,32 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
+	/// The `temporary_channel_id` parameter indicates which channel the reponse is for.
+	///
+	/// [`events::Event::OpenChannelRequest`]: crate::util::events::Event::OpenChannelRequest
+	pub fn respond_to_open_channel_request(&self, accept_channel: bool, temporary_channel_id: [u8; 32]) -> Result<(), APIError> {
```
nit: I'd kinda prefer an API of `accept_inbound_channel` with users instead calling `force_close_channel` to reject the channel, that way our code stays consistent and we use the same codepaths for everything.

ViktorTigerstrom replied:
That makes a lot of sense! I'll change to that.

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4101,6 +4101,32 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
+	pub fn respond_to_open_channel_request(&self, accept_channel: bool, temporary_channel_id: [u8; 32]) -> Result<(), APIError> {
+		let _persistence_guard = PersistenceNotifierGuard::notify_on_drop(&self.total_consistency_lock, &self.persistence_notifier);
+
+		if let Some(counterparty_node_id) = self.internal_get_counterparty_node_id(temporary_channel_id) {
```
This is fundamentally race-y - you take the lock in `internal_get_counterparty_node_id`, check if the channel is there, and then release the lock, then go into `accept_requested_channel` and take the lock again. Anyway, this will be easy to remove with the above suggestion.

ViktorTigerstrom replied:
Thanks, I'll fix it through the above suggestion.

TheBlueMatt reviewed (commented):
Basically looks good.

ViktorTigerstrom commented:
Updated the PR once more, to address the latest feedback. Also squashed all commits as they were becoming unfocused.

TheBlueMatt commented on lightning/src/util/config.rs:
```diff
@@ -305,6 +305,9 @@ pub struct UserConfig {
 	/// Default value: true.
 	pub accept_inbound_channels: bool,
+	/// If this is set to true, the user needs to manually accept inbound requests to open a new channel.
+	/// Default value: false.
```
nit: Would be good to at least link to `Event::OpenChannelRequest` here.

ViktorTigerstrom replied:
Sure, I'll add that.

TheBlueMatt reviewed (approved):
LGTM. Could really use a test (eg take a look at some of the simpler functional tests in `lightning/src/ln/functional_tests.rs`), but if you don't have time to get around to that I'll be happy to take a look.

valentinewallace commented on lightning/src/util/events.rs:
```diff
@@ -403,6 +403,37 @@ pub enum Event {
+		temporary_channel_id: [u8; 32],
+		/// The node_id of the counterparty requesting to open the channel.
+		counterparty_node_id: PublicKey,
+		/// The genesis hash of the blockchain where the channel is requested to be opened.
+		chain_hash: BlockHash,
```
I wonder if this is a useful field? It seems a bit extraneous to me

TheBlueMatt replied:
Yea, I guess not, we don't currently support multi-chain operation so the chain hash has to match what we are on anyway. ie its always the same value by the time it gets here.

ViktorTigerstrom replied:
Great ok, I'll remove the field.

valentinewallace commented on lightning/src/ln/channel.rs:
```diff
@@ -1918,6 +1933,10 @@ impl<Signer: Sign> Channel<Signer> {
 			return Err(ChannelError::Close("Received funding_created after we got the channel!".to_owned()));
 		}
+		if self.inbound_awaiting_accept {
+			// The Channel must manually accepted before the FundingCreated is received.
```
nit: must *be manually accepted

I think we can get rid of this comment though

ViktorTigerstrom replied:
I agree, I'll get rid of it.

valentinewallace commented on lightning/src/ln/channel.rs:
```diff
@@ -536,6 +536,17 @@ pub(super) struct Channel<Signer: Sign> {
 	closing_fee_limits: Option<(u64, u64)>,
 
+	/// Flag that ensures that the `get_accept_channel` function must be called before the
+	/// `funding_created` function is executed successfully. The reason for need of this flag is
```
nit: s/need of//

ViktorTigerstrom replied:
Thanks!

valentinewallace reviewed (commented):
One question about a field but otherwise ACK with testing either here or in follow-up

ViktorTigerstrom commented:
Great, thanks a lot to you both! Sure I'll look into implementing some tests. Unfortunately I've got quite much in the upcoming days, but I'll try to get it done during weekend. I could also look into implementing support for the feature in the ldk-sample repo afterwards if that's helpful.

TheBlueMatt commented:
> Great, thanks a lot to both of you! Sure I'll look into implementing some tests. Unfortunately I've got quite much in the upcoming days, but I'll try to get it done during weekend. 

No rush at all. If you have the time to address the other comments and squash the commits here, that'd be appreciated too. I'm working on 0conf support which will likely be based on this, so I may just write a test for you and we can merge it without, depending on timing.

ViktorTigerstrom commented:
> No rush at all. If you have the time to address the other comments and squash the commits here, that'd be appreciated too. I'm working on 0conf support which will likely be based on this, so I may just write a test for you and we can merge it without, depending on timing.

Ok sounds great, just pushed the suggested changes!

ViktorTigerstrom commented:
Pushed some tests of the feature if you haven't had time to create some already. I made one test for accepting an inbound request, and one for rejecting a request. 

I also wanted to make a test for the case where the counterparty node tries to be smart and send a `FundingCreated` msg before the channel has been accepted, but that proved to be more difficult to implement so I'll try to get that implemented at a later time.

I pushed the tests in a separate commit, but I can of course squash them if that's more suitable.

ViktorTigerstrom commented:
> Nice! Those tests look great. Indeed, it'd be nice to see the channel get closed if the counterparty sends a `funding_generated` (I don't think we need to worry about `funding_created`, we need `funding_generated` first) before we send the `accept_channel`.

Thanks! Great, thanks for that info.

ViktorTigerstrom commented:
Just pushed a draft of the test that rejects the msg before the channel was manually accepted. I'm not really satisfied with the solution though, so if I've misunderstood how to test the case, or if there's a simpler way to do it please let me know.

The key issue with my solution is that I need to extract the `AcceptChannel` msg that `nodes[1]` would send with the `MessageSendEvent::SendAcceptChannel` event, if the channel is accepted. This is because it's required that `nodes[0]` has called `handle_accept_channel` before the `create_funding_transaction` is called with `nodes[0]` passed. If that flow is altered, `create_funding_transaction` will not succeed. This forced me to create a new function `Channel::set_is_awaiting_accept` just for this test, that resets the `Channel::inbound_awaiting_accept` value. As the `inbound_awaiting_accept` flag is an internal private flag that controls the internal logic flow in the `Channel` struct, exposing a public function that mutates that value isn't great...

All of this is done, just for the purpose of generating the funding transaction for `nodes[0]`, and use the tx to generate the `FundingCreated` message. Obviously this is isn't necessarily required by other implementations to generate the `FundingCreated` message, but it was the only simple way I could find with our implementation.

If I've completely misunderstood the flow, or if there's any simpler way to generate the funding transaction that requires less prerequisites, I'd be really thankful for that feedback!

TheBlueMatt commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -8174,6 +8174,153 @@ fn test_override_0msat_htlc_minimum() {
+	// function, which is required in order for the create_funding_transaction function not to
+	// crash when nodes[0] is passed to it.
+	{
+		let mut channel_state_lock = nodes[1].node.channel_state.lock().unwrap();
```
Note you can use the `get_channel_ref` macro here to do most of the work.

ViktorTigerstrom replied:
Great thanks! Hadn't found that one.

TheBlueMatt commented on lightning/src/ln/channel.rs:
```diff
@@ -4504,7 +4522,15 @@ impl<Signer: Sign> Channel<Signer> {
+		self.inbound_awaiting_accept
+	}
+
+	pub fn set_is_awaiting_accept(&mut self) {
```
This is generally fine with me, actually, but we should make this `#[cfg(test)]`. An alternative way to accomplish the same test would be to split `get_accept_channel` into an utility to get the message itself and another pub method to get the message while setting the awaiting flag.

ViktorTigerstrom commented:
Thanks for the feedback! Pushed some changes to address it. I choose the alternative to split the `get_accept_channel` into a utility. Please check that we're ok with exposing a function that explicitly generates an `AcceptChannel` msg without doing sanity checks of the channel state and such. I added that functionality in the commit `6316e2a7`, but will squash it with the first commit if it looks good. 

Also added the two last test I could think of related to the manually accepting of inbound channels feature.

ViktorTigerstrom commented:
Fixed crate issue, my bad.

valentinewallace commented on lightning/src/ln/channel.rs:
```diff
@@ -4514,7 +4540,21 @@ impl<Signer: Sign> Channel<Signer> {
+
+		self.inbound_awaiting_accept = false;
+
+		return self.get_accept_channel_message();
```
nit: s/return self.get_accept_channel_message();/self.get_accept_channel_message()

valentinewallace commented on lightning/src/ln/channel.rs:
```diff
@@ -4514,7 +4540,21 @@ impl<Signer: Sign> Channel<Signer> {
+	/// [`Channel::accept_inbound_channel`] function instead.
+	///
+	/// [`msgs::AcceptChannel`]: crate::ln::msgs::AcceptChannel
+	pub fn get_accept_channel_message(&self) -> msgs::AcceptChannel {
```
Don't think it matters too much but we could make this pub just in testing

TheBlueMatt replied:
Yea, I'd strongly prefer this be a private method with a `#[cfg(test)]` method that exposes it just for testing.

ViktorTigerstrom replied:
Ok sure, I'll make it private and expose a function for testing!

valentinewallace reviewed (approved):
Testing looks super solid! Just nits here, if no further feedback I think it'd be fine to just merge

TheBlueMatt commented:
Needs rebase, sorry about that.

ViktorTigerstrom commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -8174,6 +8174,210 @@ fn test_override_0msat_htlc_minimum() {
+	// Clear the [`Event::OpenChannelRequest`] event without responding to the request.
+	nodes[1].node.get_and_clear_pending_events();
+
+	// Get the [`AcceptChannel`] message of [`nodes[1]`] without calling the
+	// [`ChannelManager::accept_inbound_channel`] function, which generates the
+	// [`MessageSendEvent::SendAcceptChannel`] event. This is passed to [`nodes[0]`]
+	// [`handle_accept_channel`] function, which is required in order for the
+	// [`create_funding_transaction`] function to succeed when [`nodes[0]`] is passed to it.
```
Oh, I was quite unsure what to do with those. Thanks a lot for clarifying!

ViktorTigerstrom commented on lightning/src/ln/channel.rs:
```diff
@@ -536,6 +536,17 @@ pub(super) struct Channel<Signer: Sign> {
 	#[cfg(not(test))]
 	closing_fee_limits: Option<(u64, u64)>,
 
+	/// Flag that ensures that the `accept_inbound_channel` function must be called before the
+	/// `funding_created` function is executed successfully. The reason for this flag is that
```
True makes sense, thanks!
